modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/GridCacheSetProxy.java

```diff
@@ -352,14 +354,26 @@ public void blockOnRemove() {
 
     /** {@inheritDoc} */
     @Override public void close() {
-        gate.enter();
+        IgniteFuture<Boolean> destroyFut = null;
+
+        if (!gate.enterIfNotStopped())
```

Comment:
do we really need to remove gate.enter(); and put  if (!gate.enterIfNotStopped()) ?